### PR TITLE
Audit 6: Disallow migration downgrades

### PR DESCRIPTION
**Pull request by adairrr**
This change adds an assertion that checks that the migration is not downgrading versions. It will return an `AbstractError::CannotDowngradeContract {...}` if a downgrade is requested.

It adds the check to 
- `ans-host`
- `manager`
- `account-factory`
- `module-factory`

These are using `cw-semver` instead of semver right now:
- `version-control`
- `ibc-client`

Because we are now testing both the entry-point messages *and* the contracts, a `test_common` to many packages was introduced that handles the `mock_init`.

### CODE REVIEW

**Comment by adairrr** [comment]
Task linked: [CU-861mkceg2 Prevent migration to older version and assert module ID](https://app.clickup.com/t/861mkceg2)

File: contracts/account/manager/src/contract.rs [hunk 1]
```diff
@@ -35,11 +36,10 @@ pub(crate) const MAX_TITLE_LENGTH: usize = 64;
...
     let version: Version = CONTRACT_VERSION.parse().unwrap();
-    let storage_version: Version = get_contract_version(deps.storage)?.version.parse().unwrap();
-    if storage_version < version {
-        set_contract_version(deps.storage, MANAGER, CONTRACT_VERSION)?;
-        migrate_module_data(deps.storage, MANAGER, CONTRACT_VERSION, None::<String>)?;
-    }
+
+    assert_contract_upgrade(deps.storage, MANAGER, version)?;
+    set_contract_version(deps.storage, MANAGER, CONTRACT_VERSION)?;
+    migrate_module_data(deps.storage, MANAGER, CONTRACT_VERSION, None::<String>)?;
```
Comment (CyberHoward) [inline contracts/account/manager/src/contract.rs:42]: Why are we adding module data to the manager, proxy, factories etc?
Imo module data is only useful in the modules (app,api) hence its name. 
it contains dependency information that is irrelevant to these contracts. The regular CW2 is good for these contracts. Less state = better

**Review by CyberHoward — CHANGES_REQUESTED** [review]
Would prefer we keep the module data to the modules (app,api).

File: contracts/account/manager/src/contract.rs (hunk 1 above)
Comment (adairrr) [reply contracts/account/manager/src/contract.rs:42]: Maybe we can have a separate `contract_metadata` or something?

Comment (CyberHoward) [reply contracts/account/manager/src/contract.rs:42]: That's something we can introduce later. Also these contracts are used purely internally so who is going to value this data? We would also need to update the instantiation and possibly allow for updates, etc. 
This is ballooning complexity without any upside for imo, hence i'd like to keep it to the modules (where there is an actual upside to storing the schema metadata)

Comment (CyberHoward) [reply contracts/account/manager/src/contract.rs:42]: If we decide to change our minds on the implementation of this data after we rolled it out we're going to have a lot of work on each of these contracts to reflect that as well.

**Review by CyberHoward — APPROVED** [review]

**Pull request merged**